ledger/src/blockstore.rs

```diff
@@ -305,50 +305,49 @@ impl Blockstore {
     /// Dangerous; Use with care:
     /// Does not check for integrity and does not update slot metas that refer to deleted slots
     /// Modifies multiple column families simultaneously
-    pub fn purge_slots(&self, mut from_slot: Slot, to_slot: Option<Slot>) {
+    pub fn purge_slots(&self, from_slot: Slot, to_slot: Option<Slot>) {
         // if there's no upper bound, split the purge request into batches of 1000 slots
         const PURGE_BATCH_SIZE: u64 = 1000;
+        let mut batch_start = from_slot;
         let mut batch_end = to_slot.unwrap_or(from_slot + PURGE_BATCH_SIZE);
         while from_slot < batch_end {
             match self.run_purge(from_slot, batch_end) {
-                Ok(end) => {
-                    if !self.no_compaction {
-                        if let Err(e) = self.compact_storage(from_slot, batch_end) {
-                            // This error is not fatal and indicates an internal error
-                            error!(
-                                "Error: {:?}; Couldn't compact storage from {:?} to {:?}",
-                                e, from_slot, batch_end
-                            );
-                        }
-                    }
-
-                    if end {
-                        break;
-                    } else {
-                        // update the next batch bounds
-                        from_slot = batch_end;
-                        batch_end = to_slot.unwrap_or(batch_end + PURGE_BATCH_SIZE);
-                    }
+                Ok(_all_columns_purged) => {
+                    // update the next batch bounds
+                    batch_start = batch_end;
+                    batch_end = to_slot.unwrap_or(batch_end + PURGE_BATCH_SIZE);
                 }
                 Err(e) => {
                     error!(
                         "Error: {:?}; Purge failed in range {:?} to {:?}",
-                        e, from_slot, batch_end
+                        e, batch_start, batch_end
                     );
                     break;
                 }
             }
         }
+
+        if !self.no_compaction {
+            if let Err(e) = self.compact_storage(from_slot, batch_end) {
```

Comment:
I'm not sure this changes anything, because this functions purge size is 1000 slots, whereas the ledger_cleanup_service batch size is 256 slots, so there will always just be a single batch for this function.

Reply:
https://github.com/solana-labs/solana/pull/10199#issuecomment-633085474